include/ibamr/INSVCStaggeredConservativeMassMomentumSSPRKIntegrator.h

```diff
@@ -0,0 +1,128 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2018 - 2020 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_INSVCStaggeredConservativeMassMomentumSSPRKIntegrator
+#define included_IBAMR_INSVCStaggeredConservativeMassMomentumSSPRKIntegrator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "ibamr/INSVCStaggeredConservativeMassMomentumRKIntegrator.h"
+
+namespace SAMRAI
+{
+namespace solv
+{
+template <int DIM, class TYPE>
+class SAMRAIVectorReal;
+template <int DIM>
+class RobinBcCoefStrategy;
+} // namespace solv
+namespace hier
+{
+template <int DIM>
+class BasePatchHierarchy;
+} // namespace hier
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class INSVCStaggeredConservativeMassMomentumSSPRKIntegrator is a derived class that integrates
+ * the staggered density field
+ *
+ *  \f$ \frac{\partial \rho}{\partial t} + \nabla \cdot (\rho u) = S(x,t) \f$
+ *
+ * and computes the conservative form of the convective operator
+ * \f$ \nabla \cdot (\rho u u)\f$.
+ *
+ * This class implements the SSP-RK2 and SSP-RK3 time stepping schemes.
+ *
+ * Class INSVCStaggeredConservativeMassMomentumSSPRKIntegrator computes the convective
+ * derivative of a side-centered velocity field using various bounded-limiters
+ * described by Patel and Natarajan.
+ *
+ * A side-centered density update is provided by this class, which is used in the
+ * conservative discretization of the incompressible Navier-Stokes equation.
+ * There is an optional mass density source term \f$ S(x,t) \f$ that can be set to check the order
+ * of accuracy via manufactured solutions.
+ *
+ * References
+ * Patel, JK. and Natarajan, G., <A HREF="https://www.sciencedirect.com/science/article/pii/S0045793014004009">
+ * A generic framework for design of interface capturing schemes for multi-fluid flows</A>
+ *
+ * \note This class is specialized in that it computes a conservative discretization of the form
+ * \f$N = \nabla \cdot (u \rho u)\f$, where the density \f$\rho\f$ can vary in space and time.
+ * This operator is to be used in conjuction with the conservative form of the variable coefficient
+ * Navier-Stokes equations, which will produce better results for high density ratio flows.
+ *
+ * \see INSVCStaggeredHierarchyIntegrator
+ */
+class INSVCStaggeredConservativeMassMomentumSSPRKIntegrator : public INSVCStaggeredConservativeMassMomentumRKIntegrator
```

Comment:
It seems odd that this is a subclass of `INSVCStaggeredConservativeMassMomentumRKIntegrator` instead of `MassIntegrator`. Are the non-SSP time steppers actually useful?

Reply:
`INSVCStaggeredConservativeMassMomentumRKIntegrator` is derived from `MassIntegrator`. SSP-RK integrator is a bit more specialized than its parent RK integrator class.

Reply:
I may be missing something, but it looks like it just reimplements `integrate`. Do we really need to have all of these different classes just to provide these simple RK schemes?

Reply:
"simple" is an understatement ;-). To maintain stability there needs to be an orchestrated dance between time stepping of INSVCHI and these mass integrators. Even a slight mismatch at the time level of these two integrators leads to instability for high density ratio cases.

Synchronizing SSP-RK with INSVC is more involved than RK2 with INSVC.

Reply:
SSP-RK2 and SSP-RK3 are essentially linear combinations of forward Euler steps. (This is one reason why it seems like you should be able to put forward Euler in this class along with the higher-order time steppers.) If the time stepping is substantially different from that, I think it would be helpful to include a few comments describing the sense in which the time stepping is using SSP-RK.

Also, there are different SSP-RK schemes --- can we document which ones are implemented?

Reply:
(BTW, documentation via pointing to papers seems fine!)

Reply:
> "simple" is an understatement ;-). To maintain stability there needs to be an orchestrated dance between time stepping of INSVCHI and these mass integrators. Even a slight mismatch at the time level of these two integrators leads to instability for high density ratio cases.
> 
> Synchronizing SSP-RK with INSVC is more involved than RK2 with INSVC.

It looks like the implementations of `integrate` only differ in 2 or 3 places --- are you sure it wouldn't be less maintenance to have some switch statements that pick the SSP vs. non-SSP code paths where they differ?